File: txmatching/web/frontend/src/app/model/Configuration.ts
import { ConfigurationGeneratedSolverConstructorNameEnum, CountryCodeGenerated } from '@app/generated';

export interface Configuration {
  use_split_resolution: boolean;
  require_compatible_blood_group: boolean;
  require_better_match_in_compatibility_index: boolean;
  require_better_match_in_compatibility_index_or_blood_group: boolean;
  use_binary_scoring: boolean;

  blood_group_compatibility_bonus: number;
  minimum_total_score: number;
  maximum_total_score: number;
  max_cycle_length: number;
  max_sequence_length: number;
  max_number_of_distinct_countries_in_round: number;
  max_matchings_to_show_to_viewer: number;
  max_number_of_matchings: number;

  // complex FE
  manual_donor_recipient_scores: DonorRecipientScore[];
  required_patient_db_ids: number[];
  forbidden_country_combinations: CountryCombination[];

  solver_constructor_name: ConfigurationGeneratedSolverConstructorNameEnum;

  // Parameters that should not be modified in FE
  max_cycles_in_all_solutions_solver: number;
  max_matchings_in_all_solutions_solver: number;
  max_matchings_in_ilp_solver: number;

  [key: string]: boolean | string | number | number[] | CountryCombination[] | DonorRecipientScore[];
}

export interface AppConfiguration extends Configuration {
  scorer_constructor_name: string;
}

export interface DonorRecipientScore {
  donor_db_id: number;
  recipient_db_id: number;
  score: number;
}

export interface CountryCombination {
  donor_country: CountryCodeGenerated;
  recipient_country: CountryCodeGenerated;
}

File: txmatching/config/configuration.py
from dataclasses import dataclass, field
from typing import List, Tuple

from txmatching.utils.country import Country

DEFAULT_FORBIDDEN_COUNTRY_LIST = [(Country.AUT, Country.IL), (Country.IL, Country.AUT)]


@dataclass
class DonorRecipientScore:
    donor_id: int
    recipient_id: int
    score: float


# pylint: disable=too-many-instance-attributes
# I think it is reasonable to have many attributes here

@dataclass
class ConfigurationBase:
    """
    Attributes:
        minimum_total_score: Minimum total score (compatibility index + blood group bonus) that is required for
        a transplant to be possible
        use_binary_scoring: If all the conditions above are satisfied, then use just 1 for possible transplant
        and -inf for impossible
    """
    scorer_constructor_name: str = "HLAAdditiveScorer"
    solver_constructor_name: str = "AllSolutionsSolver"
    enforce_compatible_blood_group: bool = False
    minimum_total_score: float = 0.0
    maximum_total_score: float = 27.0
    require_new_donor_having_better_match_in_compatibility_index: bool = False
    require_new_donor_having_better_match_in_compatibility_index_or_blood_group: bool = False
    blood_group_compatibility_bonus = 0.0
    forbidden_country_combinations: List[Tuple[Country, Country]] = field(
        default_factory=lambda: DEFAULT_FORBIDDEN_COUNTRY_LIST)
    use_binary_scoring: bool = False
    max_cycle_length: int = 100
    max_sequence_length: int = 100
    max_number_of_distinct_countries_in_round: int = 100
    required_patient_db_ids: List[int] = field(default_factory=list)
    allow_low_high_res_incompatible: bool = False


@dataclass
class Configuration(ConfigurationBase):
    """
    Attributes:
         manual_donor_recipient_scores: Manual setting of score for tuple of recipient and donor
     """
    manual_donor_recipient_scores: List[DonorRecipientScore] = field(default_factory=list)


# TODO process correctly forbidden_country_combinations and allow_low_high_res_incompatible
BOOL_KEYS_IN_CONFIG = [
    'enforce_compatible_blood_group',
    'require_new_donor_having_better_match_in_compatibility_index',
    'require_new_donor_having_better_match_in_compatibility_index_or_blood_group',
    'use_binary_scoring',
    'allow_low_high_res_incompatible'
]
FLOAT_KEYS_IN_CONFIG = [
    'minimum_total_score',
    'blood_group_compatibility_bonus'
]
INT_KEYS_IN_CONFIG = [
    'max_cycle_length',
    'max_sequence_length',
    'max_number_of_distinct_countries_in_round'

]
MAN_DON_REC_SCORES = 'manual_donor_recipient_scores'
